src/dmd/json.d

```diff
@@ -857,17 +858,165 @@ public:
 extern (C++) void json_generate(OutBuffer* buf, Modules* modules)
 {
     scope ToJsonVisitor json = new ToJsonVisitor(buf);
-    json.arrayStart();
-    json.generateCompilerInfo();
-    json.generateBuildInfo();
-    for (size_t i = 0; i < modules.dim; i++)
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQuery is null)
+    {
+        // Generate the original "non-query format, which is just an array
+        // of modules representing their syntax.
+
+        json.arrayStart();
+        for (size_t i = 0; i < modules.dim; i++)
+        {
+            Module m = (*modules)[i];
+            if (global.params.verbose)
+                message("json gen %s", m.toChars());
+            m.accept(json);
+        }
+        json.arrayEnd();
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the "query format" which is an object where each
+        // output option is it's own field.
+
+        json.objectStart();
+        auto parser = QueryParser(global.params.jsonQuery);
+        for (;;)
+        {
+            parser.skipTrivial();
+            if (*parser.next == '\0')
+                break;
+            auto fieldName = parser.tryParseName();
+            if (fieldName is null)
+            {
+                //json.generateError(format("invalid query: expected name but got '%s' (0x%02x)",
+                //    *parser.next, cast(ubyte)*parser.next));
+                json.generateError("invalid query: expected name or EOF but got something else".ptr);
```

Comment:
most useful diagnostic would be a to show byte offset from start of string where parser got confused

Reply:
(also this is uncovered)

Reply:
workin on this one

Reply:
ok, I ended up using vsnprintf, also added tests.